### PR TITLE
Connect backend and frontend for convert hearing type page

### LINKED ISSUE #15159: Connect view for converting legacy appeal travel board hearing request to virtual hearing request with backend and actions menu
Connects #14857

## Description
Connect the view/form created in #15157 with the task actions menu on the case details page created in #14899, and connect the form with the backend actions created in #15156.

## Acceptance criteria
- [ ] Selecting "Convert hearing type to virtual" on a `ChangeHearingRequestTypeTask` shows the non-modal "Convert Hearing to Virtual" form (created in #15217)
- [ ] Submitting the "Convert hearing to virtual" form causes the new request type to be recorded correctly
- [ ] The expected alert is shown when the conversion has successfully taken place
    - [ ] If the appeal has been geomatched:
    ```
    The hearing request is in the scheduling queue for the [Regional office name] regional office.
    ```
    - [ ] If the appeal has not been geomatched:
    ```
    The hearing request is in the scheduling queue for the appropriate regional office.
    ```

## Background/context/resources
[Designs and flow are here](https://www.figma.com/proto/V87TZArfdurCGJiEjQ73ES/Virtual-Hearings?node-id=6333%3A2104&viewport=-839%2C837%2C0.08404029160737991&scaling=min-zoom)

### PR DESCRIPTION
**Pull request by ferristseng**
Resolves #15159

### Description

  * Create `HearingTypeConversion` component which wraps the `HearingTypeConversionForm`
  * Create new route that's connected to the convert hearing type task action
  * Small refactors
    * Made the logic to determine whether or not to display the veteran vs. appellant name consistent
    * Attempt to consolidate where hearing request type mappings are defined in a single location

### Acceptance Criteria

- [x] Selecting "Convert hearing type to virtual" on a `ChangeHearingRequestTypeTask` shows the non-modal "Convert Hearing to Virtual" form (created in #15217)
- [x] Submitting the "Convert hearing to virtual" form causes the new request type to be recorded correctly
- [x] The expected alert is shown when the conversion has successfully taken place
    - [x] If the appeal has been geomatched:
    ```
    The hearing request is in the scheduling queue for the [Regional office name] regional office.
    ```
    - [x] If the appeal has not been geomatched:
    ```
    The hearing request is in the scheduling queue for the appropriate regional office.
    ```

### Testing Plan

N/A

### User Facing Changes

_Note: The task tree is a bit messed up because I was reusing the same appeal over and over again. When I completed the task the first time, it unlocks the schedule hearing task. From there, I was able to schedule the appeal, which is why you see the select hearing disposition task in the screencap._

Screen Recording 2020-09-28 at 10 52 01 AM mov

### Code Documentation Updates
- [x] Add or update code comments at the top of the class, module, and/or component.

### Storybook Story
*For Frontend (Presentationa) Components*
* [x] Add a [Storybook](https://github.com/department-of-veterans-affairs/caseflow/wiki/Documenting-React-Components-with-Storybook) file alongside the component file (e.g. create `MyComponent.stories.js` alongside `MyComponent.jsx`)
* [x] Give it a title that reflects the component's location within the overall Caseflow hierarchy
* [x] Write a separate story (within the same file) for each discrete variation of the component

### CODE REVIEW

File: app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -106,6 +106,8 @@ def update
     render json: {
       tasks: tasks_hash
     }
+  rescue ActiveRecord::RecordInvalid => error
+    invalid_record_error(error.record)
```
Comment (ferristseng) [inline app/controllers/tasks_controller.rb:110]: I was using this to test the error handling code on the frontend. I figured it probably makes sense to JSONify any validation errors regardless, so I kept it in the PR.

File: app/models/hearing_day.rb [hunk 2]
```diff
@@ -34,12 +34,7 @@ class HearingDay < CaseflowRecord
 
   class HearingDayHasChildrenRecords < StandardError; end
 
-  REQUEST_TYPES = {
-    virtual: "R",
-    video: "V",
-    travel: "T",
-    central: "C"
-  }.freeze
+  REQUEST_TYPES = Constants::HEARING_REQUEST_TYPES.with_indifferent_access.freeze
```
Comment (ferristseng) [inline app/models/hearing_day.rb:37]: I moved these mappings to a JSON constants file so the frontend would have access to them too.

File: client/app/hearings/components/VirtualHearings/AppellantSection.jsx [hunk 3]
```diff
@@ -28,9 +28,11 @@ export const AppellantSection = ({
   showOnlyAppellantName,
   showMissingEmailAlert
 }) => {
-  const appellantName = hearing?.appellantFullName ? hearing?.appellantFullName :
-    `${hearing?.veteranFirstName} ${hearing?.veteranLastName}`;
-
+  // Depending on where this component is used, the *FullName fields will be available.
+  // If they aren't, the *FirstName/*LastName fields should be available.
+  const appellantName = hearing?.appellantIsNotVeteran ?
```
Comment (ferristseng) [inline client/app/hearings/components/VirtualHearings/AppellantSection.jsx:33]: I made this based on the `appellantIsNotVeteran` flag, which is how we determine whether or not to use appellant vs. veteran in other places.

File: client/app/hearings/constants.js [hunk 4]
```diff
@@ -126,9 +126,3 @@ export const HEARING_CONVERSION_TYPES = [
   'change_email_or_timezone',
   'change_hearing_time'
 ];
-
-export const HEARING_REQUEST_TYPES = {
```
Comment (ferristseng) [inline client/app/hearings/constants.js:130]: I removed this and preferred directly importing `VIDEO_HEARING`, `CENTRAL_OFFICE_HEARING`, and `VIRTUAL_HEARING`. We declare these kinds of mappings in a bunch of different places, so I'm trying to consolidate where possible.

File: client/app/hearings/components/VirtualHearings/AppellantSection.jsx (hunk 3 above)
Comment (sahalliburton) [reply client/app/hearings/components/VirtualHearings/AppellantSection.jsx:33]: Thanks thats a much better way to check

**Review by sahalliburton — CHANGES_REQUESTED** [review]
LGTM, I tested both scenarios and when geomatched it does show the regional office and when not it shows the correct text. There is just one issue I noticed and that is when the hearing is converted to travel, I need to refresh the page in order to see the hearing type change to "Virtual"

**Comment by ferristseng** [comment]
> There is just one issue I noticed and that is when the hearing is converted to travel, I need to refresh the page in order to see the hearing type change to "Virtual"

It looks like this is an issue because the appeal never gets refreshed after the change is made. I updated the submit call to delete the appeal from the store, which forces the case details page to do a full reload of the appeals data.

https://github.com/department-of-veterans-affairs/caseflow/pull/15314/commits/2ac01b3031cd3a104cfa5878e6a3e986176058c5

**Review by sahalliburton — APPROVED** [review]
LGTM

**Pull request merged**